src/lib.rs

```diff
@@ -0,0 +1,560 @@
+                } else if *value < self.nodes[index].value {
+                    current_index = self.nodes[index].left_child;
+                } else {
+                    return Ok(index);
+                }
+            }
+        }
+        Err(previous_index)
+    }
+
+    fn create_node(&mut self, value: T) -> Result<Node<T>, usize> {
+        match self.get_node(&value) {
+            Ok(index) => Err(index),
+            Err(previous_index) => {
+                if let Some(parent) = previous_index {
+                    if self.nodes[parent].value < value {
+                        self.nodes[parent].right_child = Some(self.nodes.len());
+                    } else {
+                        self.nodes[parent].left_child = Some(self.nodes.len());
+                    }
+                }
+                Ok(Node {
+                    value: value,
+                    parent: previous_index,
+                    left_child: None,
+                    right_child: None,
+                    height: 0
+                })
+            }
+        }
+    }
+
+    fn update_heights(&mut self, start_index: usize) {
+        if !self.nodes.is_empty() {
+            let mut child_index = start_index;
+            while let Some(ancestor_index) = self.nodes[child_index].parent {
+                self.nodes[ancestor_index].height =
+                    self.compute_height(self.nodes[ancestor_index].left_child,
+                                        self.nodes[ancestor_index].right_child);
+                child_index = ancestor_index;
+            }
+            // update the height of the root
+            self.nodes[self.root.unwrap()].height =
+                self.compute_height(self.nodes[self.root.unwrap()].left_child,
+                                    self.nodes[self.root.unwrap()].right_child);
+        }
+    }
+
+    /// Helper function that computes the height of a node based
+    /// on the heights of its children.
+    fn compute_height(&self, l_subtree: Option<usize>,
+            r_subtree: Option<usize>) -> usize {
+        let (mut left_height, mut right_height) = (0, 0);
+        if let Some(index) = l_subtree {
+            left_height = self.nodes[index].height + 1;
+        }
+        if let Some(index) = r_subtree {
+            right_height = self.nodes[index].height + 1;
+        }
+        std::cmp::max(left_height, right_height)
+    }
+
+    fn balance_factor(&self, node: &Node<T>) -> i8 {
+        let (mut left_height, mut right_height) = (-1, -1);
+        if let Some(left_child) = node.left_child {
+            left_height = self.nodes[left_child].height as isize;
+        }
+        if let Some(right_child) = node.right_child {
+            right_height = self.nodes[right_child].height as isize;
+        }
+        (left_height - right_height) as i8
+    }
+
+    fn left_rotation(&mut self, index: usize) {
+        if let Some(right_index) = self.nodes[index].right_child {
```

Comment:
What does it mean if you don't enter the body of this conditional? i.e. if `right_index` is `None`?

Reply:
In this case, the right child of `index` becomes the parent of `index` and the left child of `right_index` becomes the right child of `index`. However, `right_index` might not have any children, which is why this condition is necessary.